File: docs/validation/substrate_math_performance/test-harness/swift/Sources/Harness/Core/VectorFile.swift
// VectorFile.swift
//
// In-memory model of a test-vector file plus JSON canonical
// serialization per `docs/test-vector-format.md`.

import Foundation

public struct VectorFile {

    public static let formatVersion = "1"
    public static let harnessVersion = "1.0.0"

    public var primitive: String
    public var cookbookSection: String
    public var generator: Generator
    public var seed: UInt64
    public var generatedAt: String
    public var outputCrc32: UInt32
    public var cases: [Case]

    public init(primitive: String,
                cookbookSection: String,
                generator: Generator,
                seed: UInt64,
                generatedAt: String,
                outputCrc32: UInt32,
                cases: [Case]) {
        self.primitive = primitive
        self.cookbookSection = cookbookSection
        self.generator = generator
        self.seed = seed
        self.generatedAt = generatedAt
        self.outputCrc32 = outputCrc32
        self.cases = cases
    }

    public struct Generator {
        public var language: String
        public var harnessVersion: String
        public var referenceFile: String

        public init(language: String, harnessVersion: String,
                    referenceFile: String) {
            self.language = language
            self.harnessVersion = harnessVersion
            self.referenceFile = referenceFile
        }
    }

    public struct Case {
        public var id: String
        public var description: String
        /// Inputs payload. Each primitive defines its own schema.
        /// Stored as canonical JSON dict (lex-sorted keys).
        public var inputs: JSONDict
        /// Expected output payload. Same shape rules as inputs.
        public var expectedOutput: JSONDict

        public init(id: String, description: String,
                    inputs: JSONDict, expectedOutput: JSONDict) {
            self.id = id
            self.description = description
            self.inputs = inputs
            self.expectedOutput = expectedOutput
        }
    }
}

/// Minimal JSON dict / value representation. Keeps key order
/// stable (sorted lex) for canonical output. Restricted to the
/// value types the test-vector format uses.
public enum JSONValue {
    case string(String)
    case integer(Int64)
    case double(Double)         // serialized as hex per format spec; NOT JSON-number
    case bool(Bool)
    case null
    case array([JSONValue])
    case dict(JSONDict)
}

public struct JSONDict {
    public private(set) var keys: [String]
    public private(set) var values: [String: JSONValue]

    public init() {
        keys = []
        values = [:]
    }

    public init(_ pairs: [(String, JSONValue)]) {
        keys = pairs.map { $0.0 }.sorted()
        values = [:]
        for (k, v) in pairs {
            values[k] = v
        }
    }

    public mutating func set(_ key: String, _ value: JSONValue) {
        if values[key] == nil {
            keys.append(key)
            keys.sort()
        }
        values[key] = value
    }

    public func get(_ key: String) -> JSONValue? {
        return values[key]
    }
}

public enum JSONWriter {

    /// Produce canonical JSON for a VectorFile per spec.
    public static func write(_ file: VectorFile) -> String {
        let topLevel = JSONDict([
            ("format_version",      .string(VectorFile.formatVersion)),
            ("primitive",           .string(file.primitive)),
            ("cookbook_section",    .string(file.cookbookSection)),
            ("generator",           .dict(JSONDict([
                ("language",         .string(file.generator.language)),
                ("harness_version",  .string(file.generator.harnessVersion)),
                ("reference_file",   .string(file.generator.referenceFile)),
            ]))),
            ("seed",                .string(HexCoding.u64(file.seed))),
            ("generated_at",        .string(file.generatedAt)),
            ("case_count",          .integer(Int64(file.cases.count))),
            ("output_crc32",        .string(HexCoding.crc32(file.outputCrc32))),
            ("cases",               .array(file.cases.map { caseToValue($0) })),
        ])
        var out = ""
        writeValue(.dict(topLevel), indent: 0, into: &out)
        out.append("\n")
        return out
    }

    private static func caseToValue(_ c: VectorFile.Case) -> JSONValue {
        return .dict(JSONDict([
            ("id",              .string(c.id)),
            ("description",     .string(c.description)),
            ("inputs",          .dict(c.inputs)),
            ("expected_output", .dict(c.expectedOutput)),
        ]))
    }

    private static func writeValue(_ v: JSONValue, indent: Int, into out: inout String) {
        let pad = String(repeating: "  ", count: indent)
        let nextPad = String(repeating: "  ", count: indent + 1)
        switch v {
        case .string(let s):
            out.append("\"")
            for ch in s {
                switch ch {
                case "\"":  out.append("\\\"")
                case "\\":  out.append("\\\\")
                case "\n":  out.append("\\n")
                case "\r":  out.append("\\r")
                case "\t":  out.append("\\t")
                default:
                    let scalar = ch.unicodeScalars.first!.value
                    if scalar < 0x20 {
                        out.append(String(format: "\\u%04x", scalar))
                    } else {
                        out.append(ch)
                    }
                }
            }
            out.append("\"")
        case .integer(let i):
            out.append(String(i))
        case .double(let d):
            // Per format spec: f64 is hex-encoded. Caller should
            // have already converted to .string(hex) before
            // reaching the writer. Reaching .double(...) here is
            // a contract violation.
            preconditionFailure("doubles must be hex-encoded as .string before reaching JSONWriter; got \(d)")
        case .bool(let b):
            out.append(b ? "true" : "false")
        case .null:
            out.append("null")
        case .array(let arr):
            if arr.isEmpty {
                out.append("[]")
            } else {
                out.append("[\n")
                for (i, item) in arr.enumerated() {
                    out.append(nextPad)
                    writeValue(item, indent: indent + 1, into: &out)
                    if i < arr.count - 1 { out.append(",") }
                    out.append("\n")
                }
                out.append(pad)
                out.append("]")
            }
        case .dict(let d):
            if d.keys.isEmpty {
                out.append("{}")
            } else {
                out.append("{\n")
                for (i, key) in d.keys.enumerated() {
                    out.append(nextPad)
                    out.append("\"\(key)\": ")
                    writeValue(d.values[key]!, indent: indent + 1, into: &out)
                    if i < d.keys.count - 1 { out.append(",") }
                    out.append("\n")
                }
                out.append(pad)
                out.append("}")
            }
        }
    }
}

public enum JSONReader {

    /// Parse a canonical-format JSON string into a JSONDict.
    /// Tolerant of extra whitespace. Hex-encoded f64 values
    /// remain strings; callers convert as needed.
    public static func parseDict(_ s: String) throws -> JSONDict {
        // For the harness we rely on Foundation's JSONSerialization
        // for parsing and then convert into our JSONDict form.
        guard let data = s.data(using: .utf8) else {
            throw JSONReaderError.invalidUTF8
        }
        let obj = try JSONSerialization.jsonObject(with: data, options: [])
        guard let dict = obj as? [String: Any] else {
            throw JSONReaderError.expectedDict
        }
        return try toJSONDict(dict)
    }

    public static func parseVectorFile(_ s: String) throws -> VectorFile {
        let root = try parseDict(s)
        guard case .string(let formatVersion) = root.get("format_version") ?? .null else {
            throw JSONReaderError.missingField("format_version")
        }
        guard formatVersion == VectorFile.formatVersion else {
            throw JSONReaderError.versionMismatch(formatVersion)
        }
        guard case .string(let primitive)        = root.get("primitive")        ?? .null,
              case .string(let cookbookSection)  = root.get("cookbook_section") ?? .null,
              case .dict(let generatorDict)      = root.get("generator")        ?? .null,
              case .string(let seedHex)          = root.get("seed")             ?? .null,
              case .string(let generatedAt)      = root.get("generated_at")     ?? .null,
              case .integer(_)                   = root.get("case_count")       ?? .null,
              case .string(let crcHex)           = root.get("output_crc32")     ?? .null,
              case .array(let caseArr)           = root.get("cases")            ?? .null else {
            throw JSONReaderError.malformedRoot
        }
        let seedBytes = try HexCoding.decode(seedHex)
        guard seedBytes.count == 8 else {
            throw JSONReaderError.malformedField("seed")
        }
        let seed = seedBytes.enumerated().reduce(UInt64(0)) { acc, item in
            acc | (UInt64(item.element) << (item.offset * 8))
        }
        let crcBytes = try HexCoding.decode(crcHex)
        guard crcBytes.count == 4 else {
            throw JSONReaderError.malformedField("output_crc32")
        }
        let crc = crcBytes.enumerated().reduce(UInt32(0)) { acc, item in
            acc | (UInt32(item.element) << (item.offset * 8))
        }
        guard case .string(let lang)        = generatorDict.get("language")         ?? .null,
              case .string(let hVer)        = generatorDict.get("harness_version")  ?? .null,
              case .string(let refFile)     = generatorDict.get("reference_file")   ?? .null else {
            throw JSONReaderError.malformedField("generator")
        }
        var cases = [VectorFile.Case]()
        for v in caseArr {
            guard case .dict(let c) = v else { throw JSONReaderError.malformedField("cases") }
            guard case .string(let id)             = c.get("id")              ?? .null,
                  case .string(let desc)           = c.get("description")     ?? .null,
                  case .dict(let inp)              = c.get("inputs")          ?? .null,
                  case .dict(let exp)              = c.get("expected_output") ?? .null else {
                throw JSONReaderError.malformedField("case")
            }
            cases.append(VectorFile.Case(id: id, description: desc,
                                          inputs: inp, expectedOutput: exp))
        }
        return VectorFile(
            primitive: primitive,
            cookbookSection: cookbookSection,
            generator: VectorFile.Generator(language: lang,
                                             harnessVersion: hVer,
                                             referenceFile: refFile),
            seed: seed,
            generatedAt: generatedAt,
            outputCrc32: crc,
            cases: cases
        )
    }

    private static func toJSONDict(_ d: [String: Any]) throws -> JSONDict {
        var pairs = [(String, JSONValue)]()
        for (k, v) in d {
            pairs.append((k, try toJSONValue(v)))
        }
        return JSONDict(pairs)
    }

    private static func toJSONValue(_ v: Any) throws -> JSONValue {
        if let s = v as? String { return .string(s) }
        if let b = v as? Bool, type(of: v) == Bool.self { return .bool(b) }
        if let n = v as? NSNumber {
            // NSNumber covers Int and Double; we discriminate.
            // Booleans are sometimes returned as NSNumber too.
            if CFGetTypeID(n) == CFBooleanGetTypeID() {
                return .bool(n.boolValue)
            }
            if String(cString: n.objCType) == "d" || String(cString: n.objCType) == "f" {
                return .double(n.doubleValue)
            }
            return .integer(n.int64Value)
        }
        if let arr = v as? [Any] {
            return .array(try arr.map { try toJSONValue($0) })
        }
        if let dict = v as? [String: Any] {
            return .dict(try toJSONDict(dict))
        }
        if v is NSNull {
            return .null
        }
        throw JSONReaderError.unsupportedValueType(String(describing: type(of: v)))
    }
}

public enum JSONReaderError: Error, CustomStringConvertible {
    case invalidUTF8
    case expectedDict
    case missingField(String)
    case malformedField(String)
    case malformedRoot
    case versionMismatch(String)
    case unsupportedValueType(String)

    public var description: String {
        switch self {
        case .invalidUTF8:              return "JSON input is not valid UTF-8"
        case .expectedDict:             return "expected top-level JSON object"
        case .missingField(let f):      return "missing required field: \(f)"
        case .malformedField(let f):    return "malformed field: \(f)"
        case .malformedRoot:            return "malformed root object"
        case .versionMismatch(let v):   return "unsupported format_version: \(v)"
        case .unsupportedValueType(let t): return "unsupported JSON value type: \(t)"
        }
    }
}

File: docs/validation/substrate_math_performance/test-harness/rust/src/harness/vector_file.rs
// src/harness/vector_file.rs
//
// In-memory model of a vector file plus canonical JSON
// (de)serialization. Mirrors the Swift harness's
// VectorFile.swift.
//
// We use `serde_json::Value` with the `preserve_order` feature so
// the JSON object key order matches what we write. Canonical JSON
// requires keys lex-sorted; we sort the keys ourselves on write.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

pub const FORMAT_VERSION: &str = "1";
pub const HARNESS_VERSION: &str = "1.0.0";

#[derive(Debug, Clone)]
pub struct VectorFile {
    pub primitive: String,
    pub cookbook_section: String,
    pub generator: Generator,
    pub seed: u64,
    pub generated_at: String,
    pub output_crc32: u32,
    pub cases: Vec<VectorCase>,
}

#[derive(Debug, Clone)]
pub struct Generator {
    pub language: String,
    pub harness_version: String,
    pub reference_file: String,
}

#[derive(Debug, Clone)]
pub struct VectorCase {
    pub id: String,
    pub description: String,
    pub inputs: JsonObject,
    pub expected_output: JsonObject,
}

/// JSON object with stable key ordering (lex-sorted on write).
pub type JsonObject = BTreeMap<String, JsonValue>;

#[derive(Debug, Clone, PartialEq)]
pub enum JsonValue {
    String(String),
    Integer(i64),
    Bool(bool),
    Null,
    Array(Vec<JsonValue>),
    Object(JsonObject),
}

// ============================================================
// Writer — canonical JSON output per spec
// ============================================================

pub struct JsonWriter;

impl JsonWriter {
    pub fn write(file: &VectorFile) -> String {
        let mut root: JsonObject = BTreeMap::new();
        root.insert("format_version".into(), JsonValue::String(FORMAT_VERSION.into()));
        root.insert("primitive".into(), JsonValue::String(file.primitive.clone()));
        root.insert(
            "cookbook_section".into(),
            JsonValue::String(file.cookbook_section.clone()),
        );

        let mut gen_obj: JsonObject = BTreeMap::new();
        gen_obj.insert(
            "language".into(),
            JsonValue::String(file.generator.language.clone()),
        );
        gen_obj.insert(
            "harness_version".into(),
            JsonValue::String(file.generator.harness_version.clone()),
        );
        gen_obj.insert(
            "reference_file".into(),
            JsonValue::String(file.generator.reference_file.clone()),
        );
        root.insert("generator".into(), JsonValue::Object(gen_obj));

        root.insert(
            "seed".into(),
            JsonValue::String(super::hex::u64_hex(file.seed)),
        );
        root.insert(
            "generated_at".into(),
            JsonValue::String(file.generated_at.clone()),
        );
        root.insert(
            "case_count".into(),
            JsonValue::Integer(file.cases.len() as i64),
        );
        root.insert(
            "output_crc32".into(),
            JsonValue::String(super::hex::u32_hex(file.output_crc32)),
        );

        let cases_arr: Vec<JsonValue> = file
            .cases
            .iter()
            .map(|c| {
                let mut o: JsonObject = BTreeMap::new();
                o.insert("id".into(), JsonValue::String(c.id.clone()));
                o.insert("description".into(), JsonValue::String(c.description.clone()));
                o.insert("inputs".into(), JsonValue::Object(c.inputs.clone()));
                o.insert(
                    "expected_output".into(),
                    JsonValue::Object(c.expected_output.clone()),
                );
                JsonValue::Object(o)
            })
            .collect();
        root.insert("cases".into(), JsonValue::Array(cases_arr));

        let mut out = String::new();
        write_value(&JsonValue::Object(root), 0, &mut out);
        out.push('\n');
        out
    }
}

fn write_value(v: &JsonValue, indent: usize, out: &mut String) {
    let pad = "  ".repeat(indent);
    let next_pad = "  ".repeat(indent + 1);
    match v {
        JsonValue::String(s) => {
            out.push('"');
            for c in s.chars() {
                match c {
                    '"' => out.push_str("\\\""),
                    '\\' => out.push_str("\\\\"),
                    '\n' => out.push_str("\\n"),
                    '\r' => out.push_str("\\r"),
                    '\t' => out.push_str("\\t"),
                    c if (c as u32) < 0x20 => {
                        out.push_str(&format!("\\u{:04x}", c as u32));
                    }
                    c => out.push(c),
                }
            }
            out.push('"');
        }
        JsonValue::Integer(i) => out.push_str(&i.to_string()),
        JsonValue::Bool(b) => out.push_str(if *b { "true" } else { "false" }),
        JsonValue::Null => out.push_str("null"),
        JsonValue::Array(arr) => {
            if arr.is_empty() {
                out.push_str("[]");
            } else {
                out.push_str("[\n");
                for (i, item) in arr.iter().enumerate() {
                    out.push_str(&next_pad);
                    write_value(item, indent + 1, out);
                    if i < arr.len() - 1 {
                        out.push(',');
                    }
                    out.push('\n');
                }
                out.push_str(&pad);
                out.push(']');
            }
        }
        JsonValue::Object(obj) => {
            if obj.is_empty() {
                out.push_str("{}");
            } else {
                out.push_str("{\n");
                let count = obj.len();
                for (i, (k, val)) in obj.iter().enumerate() {
                    out.push_str(&next_pad);
                    out.push('"');
                    out.push_str(k);
                    out.push_str("\": ");
                    write_value(val, indent + 1, out);
                    if i < count - 1 {
                        out.push(',');
                    }
                    out.push('\n');
                }
                out.push_str(&pad);
                out.push('}');
            }
        }
    }
}

// ============================================================
// Reader — parse a canonical JSON file into VectorFile
// ============================================================

pub struct JsonReader;

impl JsonReader {
    pub fn parse_vector_file(s: &str) -> Result<VectorFile, JsonReaderError> {
        let raw: serde_json::Value = serde_json::from_str(s)
            .map_err(|e| JsonReaderError::Malformed(format!("serde_json: {e}")))?;
        let root = raw
            .as_object()
            .ok_or_else(|| JsonReaderError::Malformed("root is not an object".into()))?;

        let format_version = root
            .get("format_version")
            .and_then(|v| v.as_str())
            .ok_or(JsonReaderError::MissingField("format_version"))?;
        if format_version != FORMAT_VERSION {
            return Err(JsonReaderError::VersionMismatch(format_version.to_owned()));
        }

        let primitive = root
            .get("primitive")
            .and_then(|v| v.as_str())
            .ok_or(JsonReaderError::MissingField("primitive"))?
            .to_owned();
        let cookbook_section = root
            .get("cookbook_section")
            .and_then(|v| v.as_str())
            .ok_or(JsonReaderError::MissingField("cookbook_section"))?
            .to_owned();
        let generated_at = root
            .get("generated_at")
            .and_then(|v| v.as_str())
            .ok_or(JsonReaderError::MissingField("generated_at"))?
            .to_owned();

        let seed_hex = root
            .get("seed")
            .and_then(|v| v.as_str())
            .ok_or(JsonReaderError::MissingField("seed"))?;
        let seed_bytes = super::hex::decode_hex(seed_hex)
            .map_err(|e| JsonReaderError::Malformed(format!("seed: {e}")))?;
        if seed_bytes.len() != 8 {
            return Err(JsonReaderError::Malformed("seed must be 8 bytes".into()));
        }
        let mut seed: u64 = 0;
        for (i, b) in seed_bytes.iter().enumerate() {
            seed |= (*b as u64) << (i * 8);
        }

        let crc_hex = root
            .get("output_crc32")
            .and_then(|v| v.as_str())
            .ok_or(JsonReaderError::MissingField("output_crc32"))?;
        let crc_bytes = super::hex::decode_hex(crc_hex)
            .map_err(|e| JsonReaderError::Malformed(format!("output_crc32: {e}")))?;
        if crc_bytes.len() != 4 {
            return Err(JsonReaderError::Malformed("output_crc32 must be 4 bytes".into()));
        }
        let mut crc: u32 = 0;
        for (i, b) in crc_bytes.iter().enumerate() {
            crc |= (*b as u32) << (i * 8);
        }

        let gen_obj = root
            .get("generator")
            .and_then(|v| v.as_object())
            .ok_or(JsonReaderError::MissingField("generator"))?;
        let generator = Generator {
            language: gen_obj
                .get("language")
                .and_then(|v| v.as_str())
                .ok_or(JsonReaderError::MissingField("generator.language"))?
                .to_owned(),
            harness_version: gen_obj
                .get("harness_version")
                .and_then(|v| v.as_str())
                .ok_or(JsonReaderError::MissingField("generator.harness_version"))?
                .to_owned(),
            reference_file: gen_obj
                .get("reference_file")
                .and_then(|v| v.as_str())
                .ok_or(JsonReaderError::MissingField("generator.reference_file"))?
                .to_owned(),
        };

        let cases_arr = root
            .get("cases")
            .and_then(|v| v.as_array())
            .ok_or(JsonReaderError::MissingField("cases"))?;

        let mut cases = Vec::with_capacity(cases_arr.len());
        for v in cases_arr {
            let co = v
                .as_object()
                .ok_or_else(|| JsonReaderError::Malformed("case is not an object".into()))?;
            let id = co
                .get("id")
                .and_then(|v| v.as_str())
                .ok_or(JsonReaderError::MissingField("case.id"))?
                .to_owned();
            let description = co
                .get("description")
                .and_then(|v| v.as_str())
                .ok_or(JsonReaderError::MissingField("case.description"))?
                .to_owned();
            let inputs = to_json_object(
                co.get("inputs")
                    .ok_or(JsonReaderError::MissingField("case.inputs"))?,
            )?;
            let expected_output = to_json_object(
                co.get("expected_output")
                    .ok_or(JsonReaderError::MissingField("case.expected_output"))?,
            )?;
            cases.push(VectorCase {
                id,
                description,
                inputs,
                expected_output,
            });
        }

        Ok(VectorFile {
            primitive,
            cookbook_section,
            generator,
            seed,
            generated_at,
            output_crc32: crc,
            cases,
        })
    }
}

fn to_json_object(v: &serde_json::Value) -> Result<JsonObject, JsonReaderError> {
    let obj = v
        .as_object()
        .ok_or_else(|| JsonReaderError::Malformed("expected object".into()))?;
    let mut out: JsonObject = BTreeMap::new();
    for (k, val) in obj {
        out.insert(k.clone(), to_json_value(val)?);
    }
    Ok(out)
}

fn to_json_value(v: &serde_json::Value) -> Result<JsonValue, JsonReaderError> {
    use serde_json::Value;
    match v {
        Value::String(s) => Ok(JsonValue::String(s.clone())),
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                Ok(JsonValue::Integer(i))
            } else {
                Err(JsonReaderError::Malformed(
                    "non-integer number; f64 must be hex-encoded as string".into(),
                ))
            }
        }
        Value::Bool(b) => Ok(JsonValue::Bool(*b)),
        Value::Null => Ok(JsonValue::Null),
        Value::Array(arr) => {
            let mut out = Vec::with_capacity(arr.len());
            for x in arr {
                out.push(to_json_value(x)?);
            }
            Ok(JsonValue::Array(out))
        }
        Value::Object(_) => Ok(JsonValue::Object(to_json_object(v)?)),
    }
}

#[derive(Debug, Clone)]
pub enum JsonReaderError {
    MissingField(&'static str),
    Malformed(String),
    VersionMismatch(String),
}

impl std::fmt::Display for JsonReaderError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            JsonReaderError::MissingField(s) => write!(f, "missing field: {s}"),
            JsonReaderError::Malformed(s) => write!(f, "malformed: {s}"),
            JsonReaderError::VersionMismatch(v) => write!(f, "unsupported format_version: {v}"),
        }
    }
}

impl std::error::Error for JsonReaderError {}

// serde derives left out since we serialize manually for canonical form.
// Suppress unused-import warnings via stub use:
#[allow(dead_code)]
fn _unused_serde() -> impl Serialize + for<'de> Deserialize<'de> {
    0u32
}
